internal/e2e/signals_processing_test.go

```diff
@@ -0,0 +1,292 @@
+			m.ResourceMetrics().At(0).ScopeMetrics().At(0).Metrics().Len() == 0 {
+			continue
+		}
+
+		heartbeatMetricName := "sw.otelcol.uptime"
+		generatedMetricName := "gen"
+		metricName := m.ResourceMetrics().At(0).ScopeMetrics().At(0).Metrics().At(0).Name()
+
+		switch metricName {
+		case generatedMetricName:
+			evaluateResourceAttributes(t, m.ResourceMetrics().At(0).Resource().Attributes())
+			m.ResourceMetrics().MoveAndAppendTo(gms.ResourceMetrics())
+		case heartbeatMetricName:
+			m.ResourceMetrics().MoveAndAppendTo(hbms.ResourceMetrics())
+		default:
+			continue
+		}
+	}
+	require.Equal(t, gms.MetricCount(), expectedCount)
+	evaluateHeartbeatMetric(t, hbms)
+}
+
+func evaluateTracesStream(
+	t *testing.T,
+	ctx context.Context,
+	container testcontainers.Container,
+	expectedCount int,
+) {
+	// Obtain result from container.
+	lines, err := loadResultFile(ctx, container, "/tmp/result.json")
+	require.NoError(t, err)
+
+	trs := ptrace.NewTraces()
+	ms := pmetric.NewMetrics()
+	tum := new(ptrace.JSONUnmarshaler)
+	mum := new(pmetric.JSONUnmarshaler)
+	for _, line := range lines {
+		// Traces to process.
+		tr, err := tum.UnmarshalTraces([]byte(line))
+		if err == nil && tr.ResourceSpans().Len() != 0 {
+			evaluateResourceAttributes(t, tr.ResourceSpans().At(0).Resource().Attributes())
+			tr.ResourceSpans().MoveAndAppendTo(trs.ResourceSpans())
+			continue
+		}
+
+		// Metrics to process.
+		m, err := mum.UnmarshalMetrics([]byte(line))
+		if err == nil && m.ResourceMetrics().Len() != 0 {
+			m.ResourceMetrics().MoveAndAppendTo(ms.ResourceMetrics())
+			continue
+		}
+	}
+
+	evaluateHeartbeatMetric(t, ms)
+	require.Equal(t, expectedCount, trs.SpanCount())
+}
+
+func evaluateLogsStream(
+	t *testing.T,
+	ctx context.Context,
+	container testcontainers.Container,
+	expectedCount int,
+) {
+	// Obtain result from container.
+	lines, err := loadResultFile(ctx, container, "/tmp/result.json")
+	require.NoError(t, err)
+
+	lgs := plog.NewLogs()
+	ms := pmetric.NewMetrics()
+	lum := new(plog.JSONUnmarshaler)
+	mum := new(pmetric.JSONUnmarshaler)
+	for _, line := range lines {
+		// Logs to process.
+		lg, err := lum.UnmarshalLogs([]byte(line))
+		if err == nil && lg.ResourceLogs().Len() != 0 {
+			evaluateResourceAttributes(t, lg.ResourceLogs().At(0).Resource().Attributes())
+			lg.ResourceLogs().MoveAndAppendTo(lgs.ResourceLogs())
+			continue
+		}
+
+		// Metrics to process.
+		m, err := mum.UnmarshalMetrics([]byte(line))
+		if err == nil && m.ResourceMetrics().Len() != 0 {
+			m.ResourceMetrics().MoveAndAppendTo(ms.ResourceMetrics())
+			continue
+		}
+	}
+
+	evaluateHeartbeatMetric(t, ms)
+	require.Equal(t, expectedCount, lgs.LogRecordCount())
+}
+
+func evaluateHeartbeatMetric(
+	t *testing.T,
+	ms pmetric.Metrics,
+) {
+	require.GreaterOrEqual(t, ms.ResourceMetrics().Len(), 1, "there must be at least one metric")
+	atts := ms.ResourceMetrics().At(0).Resource().Attributes()
+	v, available := atts.Get("sw.otelcol.collector.name")
```

Comment:
It could possibly utilise https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/pdatatest but I had issues with dependencies when I tried to use that.

Reply:
looks like dependencies should not be an issue in this separate package, yet I am a bit suspicious of flaky results, since the metrics count is dependent on the actual runtime (here this is mitigated by the check for at least one metric).. 
nevertheless package mentioned usage is a good idea, I'll work it into a follow-up.

Reply:
@adschr I meant something different - there were some conflicts with OTEL dependencies when I tried to use the `pdatatest` module for `heartbeat_test.go` and I couldn't compile it with that. I'm getting a different error now, though.